design/src/rfcs/rfc0023_client_crate_organization.md

```diff
@@ -0,0 +1,378 @@
+
+This RFC proposes changing the organization structure of the generated client crates to:
+
+1. Make discovery in the crate documentation easier.
+2. Facilitate re-exporting types from runtime crates in related modules without name collisions.
+
+Previous Organization
+---------------------
+
+Previously, crates were organized as such:
+
+```text
+.
+├-- client
+|   ├-- fluent_builders
+|   |   └-- <One fluent builder per operation>
+|   ├-- Builder
+|   └-- Client
+├-- config
+|   ├-- retry
+|   |   ├-- RetryConfig
+|   |   ├-- RetryConfigBuilder
+|   |   └-- RetryMode
+|   ├-- timeout
+|   |   ├-- TimeoutConfig
+|   |   └-- TimeoutConfigBuilder
+|   ├-- AsyncSleep
+|   ├-- Builder
+|   ├-- Config
+|   └-- Sleep
+├-- error
+|   ├-- <One module per error to contain a single struct named `Builder`>
+|   ├-- <One struct per error named `${error}`>
+|   ├-- <One struct per operation named `${operation}Error`>
+|   └-- <One enum per operation named `${operation}ErrorKind`>
+├-- http_body_checksum (empty)
+├-- input
+|   ├-- <One module per input to contain a single struct named `Builder`>
+|   └-- <One struct per input named `${operation}Input`>
+├-- lens (empty)
+├-- middleware
+|   └-- DefaultMiddleware
+├-- model
+|   ├-- <One module per shape to contain a single struct named `Builder`>
+|   └-- <One struct per shape>
+├-- operation
+|   ├-- customize
+|   |   ├-- ClassifyRetry
+|   |   ├-- CustomizableOperation
+|   |   ├-- Operation
+|   |   ├-- RetryKind
+|   └-- <One struct per operation>
+├-- output
+|   ├-- <One module per output to contain a single struct named `Builder`>
+|   └-- <One struct per output named `${operation}Input`>
+├-- paginator
+|   ├-- <One struct per paginated operation named `${operation}Paginator`>
+|   └-- <Zero to one struct(s) per paginated operation named `${operation}PaginatorItems`>
+├-- presigning
+|   ├-- config
+|   |   ├-- Builder
+|   |   ├-- Error
+|   |   └-- PresigningConfig
+|   └-- request
+|       └-- PresignedRequest
+├-- types
+|   ├-- AggregatedBytes
+|   ├-- Blob
+|   ├-- ByteStream
+|   ├-- DateTime
+|   └-- SdkError
+├-- AppName
+├-- Client
+├-- Config
+├-- Credentials
+├-- Endpoint
+├-- Error
+├-- ErrorExt (for some services)
+├-- PKG_VERSION
+└── Region
+```
+
+The Crate Root
+--------------
+
+The crate root should only host the most frequently used types, or phrased differently,
+the types that are critical to making a service call with default configuration,
+or that are required for the most frequent config changes (such as setting credentials,
+or changing the region/endpoint).
+
+Previously, the following were exported in root:
+```
+.
+├-- AppName
+├-- Client
+├-- Config
+├-- Credentials
+├-- Endpoint
+├-- Error
+├-- ErrorExt (for some services)
+├-- PKG_VERSION
+└── Region
+```
+
+The `AppName` is infrequently set, and should be moved into `crate::config`.
+The `Error` and `ErrorExt` types should be moved into `crate::error` since
```

Comment:
totally agree about `ErrorExt`. Depending on where you land on errors, I think keeping `Error` in the root so you can do without any duplicate `error::Error` is probably worthwhile.

The other stuff in the root like `Endpoint` should be moved. `PKG_VERSION` could be moved into a `meta` module. It's debatable if `Region` needs to be re-exported at all now that we have aws-config.

```
async fn do_a_thing() -> Result<String, aws_sdk_s3::Error> -> { ... }
```

Reply:
Your reason for keeping `Error` in the top-level makes sense to me.

I think `Region` should still be re-exported since it's referenced by `Config` and the config's `Builder`. It can be moved into `crate::config` though.